rust/macros/pinned_drop.rs

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use proc_macro::{Delimiter, Group, Ident, Punct, Spacing, Span, TokenStream, TokenTree};
+
+pub(crate) fn pinned_drop(_args: TokenStream, input: TokenStream) -> TokenStream {
+    let mut toks = input.into_iter().collect::<Vec<_>>();
+    assert!(!toks.is_empty());
+    // ensure that we have an impl item
+    assert!(matches!(&toks[0], TokenTree::Ident(i) if i.to_string() == "impl"));
+    // ensure that we are implementing `PinnedDrop`
+    let mut nesting: usize = 0;
+    let mut pinned_drop_idx = None;
+    for (i, tt) in toks.iter().enumerate() {
+        match tt {
+            TokenTree::Punct(p) if p.as_char() == '<' => {
+                nesting += 1;
+            }
+            TokenTree::Punct(p) if p.as_char() == '>' => {
+                nesting = nesting.checked_sub(1).unwrap();
+                continue;
+            }
+            _ => {}
+        }
+        if i >= 1 && nesting == 0 {
+            assert!(
+                matches!(tt, TokenTree::Ident(i) if i.to_string() == "PinnedDrop"),
+                "expected 'PinnedDrop', found: '{:?}'",
+                tt
+            );
+            pinned_drop_idx = Some(i);
+            break;
+        }
+    }
+    let idx = pinned_drop_idx.unwrap();
+    //inserting `::kernel::init::` in reverse order
```

Comment:
Maybe [`Vec::splice`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.splice) with `idx..idx` as range would work?

Reply:
https://github.com/Rust-for-Linux/linux/pull/909/commits/d1a1330fd11ea088fb49e286002da434169b1c55 didn't start using splice unlike what the commit title suggests.

Reply:
oh wrong directory in git...